Optional dependencies not listed with their corresponding extras

When running the test `tests/distribution/test_distribution_packages.py::TestExternalProject::test_optional_dependencies[pytest==7.3.0]`, setuptools-pyproject-migration does not list the corresponding extras as part of the `Requirement` objects.

I'm not sure what the correct approach should be; we can figure that out while solving this issue.

```
self = <test_distribution_packages.TestExternalProject object at 0x7f32b0ca7090>
expected = StandardMetadata(name='pytest', version=<Version('7.3.0')>, description='pytest: simple powerful testing with Python',...are Development :: Testing', 'Topic :: Utilities'], keywords=['test,unittest'], scripts={}, gui_scripts={}, dynamic=[])
actual = StandardMetadata(name='pytest', version=<Version('7.3.0')>, description='pytest: simple powerful testing with Python',..., 'unittest'], scripts={'pytest': 'pytest:console_main', 'py.test': 'pytest:console_main'}, gui_scripts={}, dynamic=[])

    def test_optional_dependencies(self, expected: StandardMetadata, actual: StandardMetadata):
        # Check that the set of extras is the same
        assert expected.optional_dependencies.keys() == actual.optional_dependencies.keys()
    
        # Check that the dependencies associated with each extra are the same
        for extra in expected.optional_dependencies:
>           assert sorted(expected.optional_dependencies[extra], key=str) == sorted(
                actual.optional_dependencies[extra], key=str
            )
E           assert [<Requirement('argcomplete; extra == "testing"')>, <Requirement('attrs>=19.2.0; extra == "testing"')>, <Requirement('hypothesis>=3.56; extra == "testing"')>, <Requirement('mock; extra == "testing"')>, <Requirement('nose; extra == "testing"')>, <Requirement('pygments>=2.7.2; extra == "testing"')>, <Requirement('requests; extra == "testing"')>, <Requirement('xmlschema; extra == "testing"')>] == [<Requirement('argcomplete')>, <Requirement('attrs>=19.2.0')>, <Requirement('hypothesis>=3.56')>, <Requirement('mock')>, <Requirement('nose')>, <Requirement('pygments>=2.7.2')>, <Requirement('requests')>, <Requirement('xmlschema')>]
E             At index 0 diff: <Requirement('argcomplete; extra == "testing"')> != <Requirement('argcomplete')>
E             Full diff:
E               [
E             -  <Requirement('argcomplete')>,
E             +  <Requirement('argcomplete; extra == "testing"')>,
E             ?                           ++++++++++++++++++++
E             -  <Requirement('attrs>=19.2.0')>,
E             +  <Requirement('attrs>=19.2.0; extra == "testing"')>,
E             ?                             ++++++++++++++++++++
E             -  <Requirement('hypothesis>=3.56')>,
E             +  <Requirement('hypothesis>=3.56; extra == "testing"')>,
E             ?                                ++++++++++++++++++++
E             -  <Requirement('mock')>,
E             -  <Requirement('nose')>,
E             +  <Requirement('mock; extra == "testing"')>,
E             +  <Requirement('nose; extra == "testing"')>,
E             -  <Requirement('pygments>=2.7.2')>,
E             +  <Requirement('pygments>=2.7.2; extra == "testing"')>,
E             ?                               ++++++++++++++++++++
E             -  <Requirement('requests')>,
E             -  <Requirement('xmlschema')>,
E             +  <Requirement('requests; extra == "testing"')>,
E             +  <Requirement('xmlschema; extra == "testing"')>,
E               ]

/home/diazona/programming/setuptools-pyproject-migration/tests/distribution/test_distribution_packages.py:146: AssertionError
```

--- a/changelog.d/134.bugfix.rst
+++ b/changelog.d/134.bugfix.rst
@@ -1,0 +1,1 @@
+Relax matching of optional dependency requirements to allow for a missing extra condition in the marker

--- a/tests/distribution/test_distribution_packages.py
+++ b/tests/distribution/test_distribution_packages.py
@@ -20,6 +20,9 @@ having to build all those wheel files, such as using already-published wheels
 from PyPI. But verifying that the procedure above works is the ultimate goal.
 """
 
+import logging
+import packaging.markers
+import packaging.requirements
 import pytest
 import sys
 
@@ -36,6 +39,9 @@ from test_support.distribution import (  # noqa: E402
     DistributionPackagePreparation,
     PyPiDistribution,
 )
+
+
+_logger = logging.getLogger("setuptools_pyproject_migration:tests:" + __name__)
 
 
 def _setuptools_scm_version_conflict() -> bool:
@@ -170,11 +176,58 @@ class TestExternalProject:
         # Check that the set of extras is the same
         assert expected.optional_dependencies.keys() == actual.optional_dependencies.keys()
 
-        # Check that the dependencies associated with each extra are the same
+        # The Requirements in the expected optional dependencies will have extra
+        # markers because those are parsed directly from core metadata's
+        # Requires-Dist lines, but in the actual optional dependencies they
+        # won't because those come from StandardMetadata.from_pyproject()
+        # which does not include extras in the Requirements. Normalizing
+        # the two to match each other is tricky because it would require
+        # parsing the markers, which is not easy; the code to do it is in
+        # the packaging package but it's not part of their public API. So for
+        # now we just compare names/urls and versions and ignore the markers.
+        def sort_key(req: packaging.requirements.Requirement):
+            return req.name, req.url, req.specifier, req.extras, str(req.marker)
+
         for extra in expected.optional_dependencies:
-            assert sorted(expected.optional_dependencies[extra], key=str) == sorted(
-                actual.optional_dependencies[extra], key=str
-            )
+            sorted_expected = sorted(expected.optional_dependencies[extra], key=sort_key)
+            sorted_actual = sorted(actual.optional_dependencies[extra], key=sort_key)
+            assert len(sorted_expected) == len(sorted_actual)
+            for e, a in zip(sorted_expected, sorted_actual):
+                assert e.name == a.name
+                assert e.url == a.url
+                assert e.specifier == a.specifier
+                assert e.extras == a.extras
+                # We expect the markers to be different because e.marker includes
+                # the extra but a.marker won't. So we look for some heuristics
+                # which are easy to detect.
+                _logger.debug("Comparing markers %r and %r", e.marker, a.marker)
+                if e.marker == a.marker:
+                    _logger.debug("Markers are equal")
+                    continue
+                elif not a.marker and e.marker == packaging.markers.Marker(f'extra == "{extra}"'):
+                    _logger.debug("Actual marker is None and expected marker is 'extra == %s'", f'"{extra}"')
+                    continue
+                # Check the same patterns that pyproject-metadata uses to add
+                # extras when writing markers out to RFC822 headers (see
+                # pyproject_metadata._build_extra_req()). Note that this isn't
+                # quite perfect; this next condition will trigger if " or "
+                # appears anywhere in the marker, even within parentheses,
+                # whereas the implementation of _build_extra_req() (correctly)
+                # only looks for an "or" token at the top level of the marker,
+                # but if that becomes an issue we'll hopefully know because
+                # a test will fail, and at that time we can make this check
+                # smarter.
+                elif " or " in str(a.marker):
+                    if e.marker == packaging.markers.Marker(f'({a.marker}) and extra == "{extra}"'):
+                        _logger.debug("Expected marker matches '(actual marker) and extra == %s'", f'"{extra}"')
+                        continue
+                else:
+                    if e.marker == packaging.markers.Marker(f'{a.marker} and extra == "{extra}"'):
+                        _logger.debug("Expected marker matches 'actual marker and extra == %s'", f'"{extra}"')
+                        continue
+                # We know this will fail but writing it as an assert rather than
+                # a direct call to pytest.fail() gives a better error message
+                assert e.marker == a.marker
 
     # Not part of core metadata
     # def test_entrypoints(self, expected: StandardMetadata, actual: StandardMetadata):

--- a/tests/distribution/test_distribution_packages.py
+++ b/tests/distribution/test_distribution_packages.py
@@ -72,7 +72,6 @@ distributions: List = [
         marks=[
             pytest.mark.distribute(
                 {
-                    "test_optional_dependencies": pytest.mark.xfail(reason="Issue #134"),
                     "test_authors": pytest.mark.xfail(reason="Issue #135"),
                 }
             ),
@@ -84,7 +83,6 @@ distributions: List = [
         marks=pytest.mark.distribute(
             {
                 "test_readme": pytest.mark.xfail,
-                "test_optional_dependencies": pytest.mark.xfail(reason="Issue #134"),
             }
         ),
     ),

